test/lib/jdk/test/lib/hotspot/ir_framework/TestFramework.java

```diff
@@ -0,0 +1,1226 @@
+ * The IR regex matching is done on the output of {@code -XX:+PrintIdeal} and {@code -XX:+PrintOptoAssembly} by parsing
+ * the hotspot_pid file of the test VM. Failing IR rules are reported by throwing a {@link IRViolationException}.
+ *
+ * @see Test
+ * @see Check
+ * @see Run
+ * @see IR
+ */
+public class TestFramework {
+    /**
+     * JTreg can define additional VM (-Dtest.vm.opts) and Javaoptions (-Dtest.java.opts) flags. IR verification is only
+     * performed when all these additional JTreg flags (does not include additionally added framework and scenario flags
+     * by user code) are whitelisted.
+     *
+     * A flag is whitelisted if it is a property flag (starting with -D), -ea, -esa, or if the flag name contains any of
+     * the entries of this list as a substring (partial match).
+     */
+    public static final Set<String> JTREG_WHITELIST_FLAGS = new HashSet<>(
+            Arrays.asList(
+                    // The following substrings are part of more than one VM flag
+                    "RAM",
+                    "Heap",
+                    "Trace",
+                    "Print",
+                    "Verify",
+                    "TLAB",
+                    "UseNewCode",
+                    // The following substrings are only part of one VM flag (=exact match)
+                    "CreateCoredumpOnCrash",
+                    "IgnoreUnrecognizedVMOptions",
+                    "UnlockDiagnosticVMOptions",
+                    "UnlockExperimentalVMOptions",
+                    "BackgroundCompilation",
+                    "Xbatch",
+                    "TieredCompilation"
+            )
+    );
+
+    static final boolean VERBOSE = Boolean.getBoolean("Verbose");
+    static final boolean TESTLIST = !System.getProperty("Test", "").isEmpty();
+    static final boolean EXCLUDELIST = !System.getProperty("Exclude", "").isEmpty();
+    static final String TEST_VM_FLAGS_START = "##### TestFrameworkPrepareFlags - used by TestFramework #####";
+    static final String TEST_VM_FLAGS_DELIMITER = " ";
+    static final String TEST_VM_FLAGS_END = "----- END -----";
+
+    private static final int WARMUP_ITERATIONS = Integer.getInteger("Warmup", -1);
+    private static final boolean PREFER_COMMAND_LINE_FLAGS = Boolean.getBoolean("PreferCommandLineFlags");
+    private static final boolean EXCLUDE_RANDOM = Boolean.getBoolean("ExcludeRandom");
+    private static final boolean REPORT_STDOUT = Boolean.getBoolean("ReportStdout");
+    private final boolean VERIFY_VM = Boolean.getBoolean("VerifyVM") && Platform.isDebugBuild();
+    private boolean VERIFY_IR = Boolean.parseBoolean(System.getProperty("VerifyIR", "true"));
+    private boolean shouldVerifyIR; // Should we perform IR matching?
+    private static String lastTestVMOutput;
+
+    private final Class<?> testClass;
+    private List<Class<?>> helperClasses = null;
+    private List<Scenario> scenarios = null;
+    private final List<String> flags = new ArrayList<>();
```

Comment:
why did you decide to eagerly create list for `flags`, but not for `scenarios` / `helpersClasses`?